lib/semantics/io.cc

```diff
@@ -0,0 +1,619 @@
+        charConst{evaluate::UnwrapExpr<DefaultCharConstantType>(foldExpr)}) {
+      hasAsynchronousYes_ = parser::ToUpperCaseLetters(**charConst) == "YES"s;
+      CheckStringValue(SpecifierKind::Asynchronous, **charConst,
+          parser::FindSourceLocation(spec));  // C1223
+    }
+  }
+}
+
+void IoChecker::Enter(const parser::IoControlSpec::CharExpr &spec) {
+  SpecifierKind specKind{};
+  using ParseKind = parser::IoControlSpec::CharExpr::Kind;
+  switch (std::get<ParseKind>(spec.t)) {
+  case ParseKind::Advance: specKind = SpecifierKind::Advance; break;
+  case ParseKind::Blank: specKind = SpecifierKind::Blank; break;
+  case ParseKind::Decimal: specKind = SpecifierKind::Decimal; break;
+  case ParseKind::Delim: specKind = SpecifierKind::Delim; break;
+  case ParseKind::Pad: specKind = SpecifierKind::Pad; break;
+  case ParseKind::Round: specKind = SpecifierKind::Round; break;
+  case ParseKind::Sign: specKind = SpecifierKind::Sign; break;
+  }
+  SetSpecifier(specKind);
+  if (const auto *expr{GetExpr(std::get<parser::ScalarDefaultCharExpr>(spec.t)
+                                   .thing.thing.value())}) {
+    const auto &foldExpr{
+        evaluate::Fold(context_.foldingContext(), common::Clone(*expr))};
+    if (const DefaultCharConstantType *
+        charConst{evaluate::UnwrapExpr<DefaultCharConstantType>(foldExpr)}) {
+      if (specKind == SpecifierKind::Advance) {
+        hasAdvanceYes_ = parser::ToUpperCaseLetters(**charConst) == "YES"s;
+      }
+      CheckStringValue(specKind, **charConst, parser::FindSourceLocation(spec));
+    }
+  }
+}
+
+void IoChecker::Enter(const parser::IoControlSpec::Pos &spec) {
+  SetSpecifier(SpecifierKind::Pos);
+}
+
+void IoChecker::Enter(const parser::IoControlSpec::Rec &spec) {
+  SetSpecifier(SpecifierKind::Rec);
+}
+
+void IoChecker::Enter(const parser::IoControlSpec::Size &spec) {
+  SetSpecifier(SpecifierKind::Size);
+}
+
+void IoChecker::Enter(const parser::IoUnit &spec) {
+  if (const parser::Variable * var{std::get_if<parser::Variable>(&spec.u)}) {
+    // TODO: C1201 - internal file variable must not be an array section
+    auto expr{AnalyzeExpr(context_, *var)};
+    if (expr && expr->GetType()->kind != 1) {
+      // This may be too restrictive; other kinds may be valid.
+      context_.Say("invalid internal file variable type"_err_en_US);  // C1202
+    }
+    SetSpecifier(SpecifierKind::Unit);
+    hasInternalUnit_ = true;
+  } else if (std::get_if<parser::Star>(&spec.u)) {
+    SetSpecifier(SpecifierKind::Unit);
+    hasStarUnit_ = true;
+  }
+}
+
+void IoChecker::Enter(const parser::MsgVariable &spec) {
+  SetSpecifier(SpecifierKind::Iomsg);
+}
+
+void IoChecker::Enter(const parser::OutputItem &spec) {
+  hasDataList_ = true;
+  // TODO: C1233 - output item must not be a procedure pointer
+}
+
+void IoChecker::Enter(const parser::StatusExpr &spec) {
+  SetSpecifier(SpecifierKind::Status);
+  if (const auto *expr{GetExpr(spec)}) {
+    const auto &foldExpr{
```

Comment:
`Fold` doesn't return a reference, const or otherwise; it returns a value.  Just use `auto foldExpr{...}`.  I'm surprised this compiled, and maybe it's a memory leak as written.

Reply:
Done. (The single remaining instance of this code is now in check-io.h.)